src/autumnaton.ts

```diff
@@ -0,0 +1,189 @@
+  "rightarm1",
+  "rightleg1",
+  "cowcatcher",
+  "periscope",
+  "radardish",
+] as const;
+
+export function mostValuableUpgrade(fullLocations: Location[]): Location[] {
+  // This function shouldn't be getting called if we don't have an expedition left
+  if (expectedRemainingExpeditions() < 1) {
+    return fullLocations;
+  }
+  const currentUpgrades = AutumnAton.currentUpgrades();
+  const acquirableUpgrades = profitRelevantUpgrades.filter(
+    (upgrade) => !currentUpgrades.includes(upgrade)
+  );
+
+  if (acquirableUpgrades.length === 0) {
+    return fullLocations;
+  }
+
+  const currentBestLocation = maxBy(fullLocations, (loc: Location) => averageAutumnatonValue(loc));
+  const currentExpectedProfit =
+    averageAutumnatonValue(currentBestLocation) * expectedRemainingExpeditions();
+
+  const upgradeValuations = acquirableUpgrades.map((upgrade) => {
+    const upgradeLocations = fullLocations.filter(
+      (location) => AutumnAton.getUniques(location)?.upgrade === upgrade
+    );
+    const bestLocContainingUpg = maxBy(upgradeLocations, (loc: Location) =>
+      averageAutumnatonValue(loc)
+    );
+
+    if (upgrade === ("periscope" || "radardish")) {
+      const bestLocWithInstalledUpg = maxBy(fullLocations, (loc: Location) =>
+        averageAutumnatonValue(loc, AutumnAton.visualAcuity() + 1)
+      );
+      const extraExpectedProfit =
+        averageAutumnatonValue(bestLocContainingUpg) +
+        averageAutumnatonValue(bestLocWithInstalledUpg) *
+          Math.max(0, expectedRemainingExpeditions() - 1);
+
+      return { upgrade, profit: extraExpectedProfit };
+    }
+    if (upgrade === ("rightleg1" || "leftleg1")) {
+      const bestLocWithInstalledUpg = currentBestLocation;
+      const extraExpectedProfit =
+        averageAutumnatonValue(bestLocContainingUpg) +
+        averageAutumnatonValue(bestLocWithInstalledUpg) *
+          Math.max(0, expectedRemainingExpeditions(AutumnAton.legs() + 1) - 1);
+
+      return { upgrade, profit: extraExpectedProfit };
+    }
+    if (upgrade === ("rightarm1" || "leftarm1")) {
+      const bestLocWithInstalledUpg = maxBy(fullLocations, (loc: Location) =>
+        averageAutumnatonValue(loc, undefined, AutumnAton.zoneItems() + 1)
+      );
+      const extraExpectedProfit =
+        averageAutumnatonValue(bestLocContainingUpg) +
+        averageAutumnatonValue(bestLocWithInstalledUpg) *
+          Math.max(0, expectedRemainingExpeditions() - 1);
+
+      return { upgrade, profit: extraExpectedProfit };
+    }
+    if (upgrade === "cowcatcher") {
+      const bestLocWithInstalledUpg = currentBestLocation;
+      const extraExpectedProfit =
+        averageAutumnatonValue(bestLocContainingUpg) +
+        (seasonalItemValue(bestLocWithInstalledUpg) +
+          averageAutumnatonValue(bestLocWithInstalledUpg)) *
+          Math.max(0, expectedRemainingExpeditions() - 1);
+
+      return { upgrade, profit: extraExpectedProfit };
+    }
+    return { upgrade, profit: 0 };
+  });
```

Comment:
Oh that function is even bigger than I thought. We might want to chunk it further once we pull it outside of the nested function.